### PR TITLE
run-enclave: Verify signing certificate

### PR DESCRIPTION
**Pull request by raulmldv**
### Verify signing certificate before loading the image

To avoid ambiguous errors for invalid signed EIFs, `run-enclave` command extracts the signature section from the image and performs certificate validation. It checks the .pem format of the certificate, the signature against the public key and if the certificate is expired.

Suggestive error messages along with a new error code have been added to make the output more explicit.

Output:

```
$ ./nitro-cli run-enclave --cpu-count 2 --memory 256 --eif-path hello-expired.eif 
[ E59 ] Signature checker error. Such error appears when validation of the signing certificate fails.

For more details, please visit https://docs.aws.amazon.com/enclaves/latest/user/cli-errors.html#E59

If you open a support ticket, please provide the error log found at "/var/log/nitro_enclaves/err2021-09-24T09:12:45.367481952+00:00.log"
Failed connections: 1
[ E39 ] Enclave process connection failure. Such error appears when the enclave manager fails to connect to at least one enclave process for retrieving the description information.

For more details, please visit https://docs.aws.amazon.com/enclaves/latest/user/cli-errors.html#E39

If you open a support ticket, please provide the error log found at "/var/log/nitro_enclaves/err2021-09-24T09:12:45.367624877+00:00.log"

$ cat /var/log/nitro_enclaves/err2021-09-24T09:12:45.367481952+00:00.log
  Action: Run Enclave
  Subactions:
    Failed to execute command `Run`
    Failed to trigger enclave run
    Invalid signing certificate: "The signing certificate is expired"
  Root error file: src/enclave_proc/commands.rs
  Root error line: 83
  Build commit: v1.0.12-38-ga3476c6-dirty
```

Issue #243 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: eif_utils/src/lib.rs [hunk 1]
```diff
@@ -611,6 +613,126 @@ impl SignCertificateInfo {
...
+        eif_file
+            .seek(SeekFrom::Start(curr_seek as u64))
+            .map_err(|e| format!("Failed to seek file from start: {:?}", e))?;
+
+        let mut section_buf = vec![0u8; EifSectionHeader::size()];
+
+        // Read all section headers and skip if different from signature section
+        while eif_file
+            .read_exact(&mut section_buf)
+            .map_err(|e| format!("Error while reading EIF header: {:?}", e))
```
Comment (andraprs) [inline eif_utils/src/lib.rs:658]: "reading EIF section header"

File: eif_utils/src/lib.rs [hunk 2]
```diff
@@ -611,6 +613,126 @@ impl SignCertificateInfo {
...
+        {
+            let section = EifSectionHeader::from_be_bytes(&section_buf)
+                .map_err(|e| format!("Error extracting EIF section header: {:?}", e))?;
+            curr_seek += EifSectionHeader::size();
+
+            if section.section_type == EifSectionType::EifSectionSignature {
+                let mut buf = vec![0u8; section.section_size as usize];
+                eif_file
+                    .seek(SeekFrom::Start(curr_seek as u64))
+                    .map_err(|e| format!("Failed to seek after EIF header: {:?}", e))?;
```
Comment (andraprs) [inline eif_utils/src/lib.rs:669]: "after EIF section header"

File: eif_utils/src/lib.rs [hunk 3]
```diff
@@ -611,6 +613,126 @@ impl SignCertificateInfo {
...
+                eif_file
+                    .read_exact(&mut buf)
+                    .map_err(|e| format!("Error while reading kernel from EIF: {:?}", e))?;
```
Comment (andraprs) [inline eif_utils/src/lib.rs:672]: "signature section" instead of "kernel"?

File: eif_utils/src/lib.rs [hunk 4]
```diff
@@ -611,6 +613,126 @@ impl SignCertificateInfo {
...
+
+                // Deserialize PCR0 signature structure and save certificate and signature
```
Comment (andraprs) [inline eif_utils/src/lib.rs:674]: "PCR signature structure" instead of "PCR0 signature structure"?

File: src/common/document_errors.rs [hunk 5]
```diff
@@ -67,6 +67,7 @@ lazy_static! {
             (NitroCliErrorEnum::LoggerError, "E56"),
             (NitroCliErrorEnum::HasherError, "E57"),
             (NitroCliErrorEnum::EnclaveNamingError, "E58"),
+            (NitroCliErrorEnum::SignatureCheckerError, "E59"),
```
Comment (andraprs) [inline src/common/document_errors.rs:70]: Could use here: "EIFSignatureCheckerError", to be more specific which signature.

File: src/common/mod.rs [hunk 6]
```diff
@@ -176,6 +176,8 @@ pub enum NitroCliErrorEnum {
     HasherError,
     /// Enclave naming error
     EnclaveNamingError,
+    /// Signature checker error
+    SignatureCheckerError,
```
Comment (andraprs) [inline src/common/mod.rs:180]: Similar as above.

File: src/enclave_proc/commands.rs [hunk 7]
```diff
@@ -71,6 +71,22 @@ pub fn run_enclaves(
         e.add_subaction("Failed to construct EnclaveManager with given arguments".to_string())
     })?;
 
+    let mut signature_checker = PcrSignatureChecker::from_eif(&args.eif_path).map_err(|e| {
+        new_nitro_cli_failure!(
+            &format!("Failed to create signature checker: {:?}", e),
```
Comment (andraprs) [inline src/enclave_proc/commands.rs:76]: Could use "create EIF signature checker".

**Review by andraprs — APPROVED** [review]

File: eif_utils/src/lib.rs [hunk 8]
```diff
@@ -611,6 +613,126 @@ impl SignCertificateInfo {
...
+        eif_file
+            .seek(SeekFrom::Start(curr_seek as u64))
+            .map_err(|e| format!("Failed to seek file from start: {:?}", e))?;
+
+        let mut section_buf = vec![0u8; EifSectionHeader::size()];
+
+        // Read all section headers and skip if different from signature section
+        while eif_file
+            .read_exact(&mut section_buf)
+            .map_err(|e| format!("Error while reading EIF section header: {:?}", e))
```
Comment (costisin) [inline eif_utils/src/lib.rs:658]: is this error conversion necessary if you don't use this error in your code?

File: eif_utils/src/lib.rs [hunk 9]
```diff
@@ -611,6 +613,126 @@ impl SignCertificateInfo {
...
+                let mut buf = vec![0u8; section.section_size as usize];
+                eif_file
+                    .seek(SeekFrom::Start(curr_seek as u64))
+                    .map_err(|e| format!("Failed to seek after EIF section header: {:?}", e))?;
+                eif_file.read_exact(&mut buf).map_err(|e| {
+                    format!("Error while reading signature section from EIF: {:?}", e)
+                })?;
+
+                // Deserialize PCR signature structure and save certificate and signature
+                let des_sign: Vec<PcrSignature> = from_slice(&buf[..])
```
Comment (costisin) [inline eif_utils/src/lib.rs:675]: Aren't these 2 equivalent?

```suggestion
                let des_sign: Vec<PcrSignature> = from_slice(&buf)
```

File: eif_utils/src/lib.rs [hunk 10]
```diff
@@ -611,6 +613,126 @@ impl SignCertificateInfo {
...
+        })
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.signing_certificate.len() == 0 && self.signature.len() == 0
+    }
+
+    /// Verifies the validity of the signing certificate
+    pub fn verify(&mut self) -> Result<(), String> {
+        let signature = COSESign1::from_bytes(&self.signature[..])
```
Comment (costisin) [inline eif_utils/src/lib.rs:700]:
```suggestion
        let signature = COSESign1::from_bytes(&self.signature)
```

File: eif_utils/src/lib.rs [hunk 11]
```diff
@@ -611,6 +613,126 @@ impl SignCertificateInfo {
...
+            .map_err(|err| format!("Could not deserialize the signature: {:?}", err))?;
+        let cert = openssl::x509::X509::from_pem(&self.signing_certificate[..])
```
Comment (costisin) [inline eif_utils/src/lib.rs:702]:
```suggestion
        let cert = openssl::x509::X509::from_pem(&self.signing_certificate)
```

File: src/enclave_proc/commands.rs [hunk 12]
```diff
@@ -71,6 +71,22 @@ pub fn run_enclaves(
         e.add_subaction("Failed to construct EnclaveManager with given arguments".to_string())
     })?;
 
+    let mut signature_checker = PcrSignatureChecker::from_eif(&args.eif_path).map_err(|e| {
+        new_nitro_cli_failure!(
+            &format!("Failed to create EIF signature checker: {:?}", e),
+            NitroCliErrorEnum::EIFSignatureCheckerError
+        )
+    })?;
+    // Verify the certificate only if signature section exists
```
Comment (costisin) [inline src/enclave_proc/commands.rs:80]:
```suggestion
    })?;
    
    // Verify the certificate only if signature section exists
```

File: eif_utils/src/lib.rs (hunk 8 above)
Comment (raulmldv) [reply eif_utils/src/lib.rs:658]: It's not necessary, I'll remove it.

File: src/enclave_proc/commands.rs (hunk 7 above)
Comment (raulmldv) [reply src/enclave_proc/commands.rs]: Followed all suggestions and modified.

**Review by andraprs — DISMISSED** [review]

**Review by costisin — DISMISSED** [review]

**Review by axlprv — APPROVED** [review]

**Review by andraprs — APPROVED** [review]

**Review by costisin — APPROVED** [review]

**Review by andraprs — APPROVED** [review]

**Review by axlprv — APPROVED** [review]

**Pull request merged**